src/Widget/BaseWidget.php

```diff
@@ -0,0 +1,162 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Bolt\Widget;
+
+use Bolt\Snippet\Target;
+use Bolt\Snippet\Zone;
+use Cocur\Slugify\Slugify;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Twig\Environment;
+
+class BaseWidget
+{
+    protected $name = 'Nameless widget';
+    protected $type = 'widget';
+    protected $target = Target::NOWHERE;
+    protected $zone = Zone::EVERYWHERE;
+    protected $priority = 0;
+    protected $context = [];
+
+    /** @var Environment */
+    protected $twig;
+
+    /** @var string */
+    protected $template;
+
+    /** @var Request */
+    protected $request;
+
+    /** @var Response */
+    protected $response;
+
+    public function setName(string $name): self
+    {
+        $this->name = $name;
+
+        return $this;
+    }
+
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    public function setType(string $type): self
+    {
+        $this->type = $type;
+
+        return $this;
+    }
+
+    public function getType(): string
+    {
+        return $this->type;
+    }
+
+    public function setTarget(string $target): self
+    {
+        $this->target = $target;
+
+        return $this;
+    }
+
+    public function getTarget(): string
+    {
+        return $this->target;
+    }
+
+    public function setPriority(int $priority): self
+    {
+        $this->priority = $priority;
+
+        return $this;
+    }
+
+    public function getPriority(): int
+    {
+        return $this->priority;
+    }
+
+    public function setTwig(Environment $twig): self
+    {
+        $this->twig = $twig;
+
+        return $this;
+    }
+
+    public function invoke(?string $template = null): string
```

Comment:
how can it use different template that has been set for it? I don't see a place where it is used.